### PR TITLE
PR: More autoindentation fixes

### LINKED ISSUE #4074: Identation broken after statements with closing parenthesis followed by more code
## Description of your problem

Indentation is broken, if a line contains a closing parenthesis followed by more code. 

Additionally, comments seem to break the behaviour of `<enter>` auto-indentation as well.

## Reproducible examples

In the following screenshot the "weird indentation" was produced by automatic indentation when pressing "Enter":

spyder-indent-1

## Versions and main components

* Spyder Version: 3.1.2
* Python Version: `3.5.2 |Anaconda 4.2.0 (64-bit)| (default, Jul  5 2016, 11:41:13) [MSC v.1900 64 bit (AMD64)]`
* Operating system: Windows 10 Professional, 64bit

## Dependencies

```
jedi >=0.8.1     :  0.9.0 (OK)
matplotlib >=1.0 :  1.5.3 (OK)
nbconvert >=4.0  :  5.1.1 (OK)
numpy >=1.7      :  1.11.1 (OK)
pandas >=0.13.1  :  0.18.1 (OK)
pep8 >=0.6       :  1.7.0 (OK)
pyflakes >=0.6.0 :  1.5.0 (OK)
pygments >=2.0   :  2.2.0 (OK)
pylint >=0.25    :  1.6.5 (OK)
qtconsole >=4.2.0:  4.2.1 (OK)
rope >=0.9.4     :  0.9.4-1 (OK)
sphinx >=0.6.6   :  1.4.6 (OK)
sympy >=0.7.3    :  1.0 (OK)
```

### LINKED ISSUE #4096: More issues with autoindentation
This was reported in StackOverflow

http://stackoverflow.com/questions/41937371/simple-auto-indentation-mode-in-spyder-ide

@rlaverde, please work on this one.

### PR DESCRIPTION
**Pull request by rlaverde**
Fixes #4096 
Fixes #4074 

I fixed some issues reported applying hanging indent for long text only

```
a = {
     } #not hanging indent
```
```
some_long_name = {
              } #hanging indent (8 spaces)
```

### CODE REVIEW

File: spyder/widgets/sourcecode/codeeditor.py [hunk 1]
```diff
@@ -2698,10 +2698,9 @@ def keyPressEvent(self, event):
                    and self.codecompletion_enter:
                     self.select_completion_list()
                 else:
-                    cmt_or_str = self.in_comment_or_string()
                     self.textCursor().beginEditBlock()
                     TextEditBaseWidget.keyPressEvent(self, event)
-                    self.fix_indent(comment_or_string=cmt_or_str)
```
Comment (ccordoba12) [inline spyder/widgets/sourcecode/codeeditor.py]: The problem with deactivating this is that indentation would become active for when you're working inside block comments. That's why I added this.

Instead we should be smarter and check if the line *doesn't* start with a comment and *ends* with it one to apply indentation in that case.

Comment (rlaverde) [inline spyder/widgets/sourcecode/codeeditor.py]: Oh I see, then I'll add an extra check for detecting inline comments.

**Comment by rlaverde** [comment]
This is ready

Should we open a new issue for `simple indentation` as mentioned in #4096? I think It could be an useful feature for some people, and also when working with non-python files

**Comment by ccordoba12** [comment]
> Should we open a new issue for simple indentation ... ?

Yes, please do it, assign it to yourself and set it for 3.2.

----

Don't worry about the test failures. I'm working hard to make them more reliable (I'm almost there ;-)

File: spyder/widgets/sourcecode/codeeditor.py [hunk 2]
```diff
@@ -1979,15 +1979,43 @@ def fix_indent(self, forward=True, comment_or_string=False):
                  (prevtext.strip().startswith("return") and
                   len(re.split(r'\(|\{|\[', prevtext)) ==
                   len(re.split(r'\)|\}|\]', prevtext)))):
+                print("hola2")
```
Comment (ccordoba12) [inline spyder/widgets/sourcecode/codeeditor.py]: Remove this print

Comment (rlaverde) [inline spyder/widgets/sourcecode/codeeditor.py]: :see_no_evil:

**Comment by ccordoba12** [comment]
Great work @rlaverde! This seems to solve all autoindentation issues reported by our users in 3.1 :-)

I simply pushed a commit to add some comments and fix column width of some lines.

File: spyder/widgets/sourcecode/tests/test_autoindent.py [hunk 3]
```diff
@@ -153,6 +153,14 @@ def test_first_line():
...
         ("s = a[(a['a'] == l) & (a['a'] == 1)]['a']\n", "s = a[(a['a'] == l) & (a['a'] == 1)]['a']\n",
          "test_balanced_brackets"),
+        ("a = (a  #  some comment\n", "a = (a  #  some comment\n     ", "test_inline_comment"),
+        ("len(a) == 1\n", "len(a) == 1\n", "test_balanced_brackets_not_ending_in_bracket"),
+        ("x = f(\n", "x = f(\n      ", "test_short_open_bracket_not_hanging_indent"),
+
+        pytest.mark.xfail(
+            ("foo = 1  # Comment open parenthesis (\n",
+             "foo = 1  # Comment open parenthesis (\n",
+             "test_comment_with parenthesis")),
```
Comment (ccordoba12) [inline spyder/widgets/sourcecode/tests/test_autoindent.py]: Could we solve this one for 3.1.4?

Comment (ccordoba12) [inline spyder/widgets/sourcecode/tests/test_autoindent.py]: I mean, it doesn't seem that hard to address :-)

Comment (rlaverde) [inline spyder/widgets/sourcecode/tests/test_autoindent.py]: Yes, although this case of use is very rare, I was thinking address it in the rewrite for 3.2

Comment (ccordoba12) [inline spyder/widgets/sourcecode/tests/test_autoindent.py]: Ok, even better :-)

**Pull request merged**